lib/types.nix

```
@@ -438,10 +447,32 @@ rec {
         description = freeformType.description or name;
         check = x: isAttrs x || isFunction x || path.check x;
         merge = loc: defs:
+          let
+            # Every attrsOf-like option adds an entry to the loc list indicating
+            # that its value isn't a defined option, but rather an arbitrary
+            # name. We pass these to the submodule such that it can know its
+            # context. We reverse this list such that static list indices can be
+            # used to get a specific attrs context.
+            # Example: For an option with a type like
+            #   attrsOf (attrsOf (submodule ..))
+            # and a definition of
+            #   { foo.bar = {}; }
+            # nameStack would be
+            #   [ "bar" "foo" .. ]
+            # where .. can be more entries depending on where the submodule
+            # is evaluated
+            nameStack = reverseList (map (l: l.name) (filter (l: l.type or "" == "attrs") loc));
+          in
           (evalModules {
             modules = allModules defs;
             inherit specialArgs;
-            args.name = last loc;
+            args = {
+              inherit nameStack;
+            } // optionalAttrs (nameStack != []) {
+              # For backwards compatibility, expose the most relevant name
+              # directly
```

```suggestion
              # Conveniently expose the most relevant name directly
```

This is the most used one, so let's keep it around.